activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,34 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def validation_needed?(klass, record, attribute)
+        return true if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        attributes = scope + [attribute.to_s]
+
+        return true if attributes.any? { |attribute| record.class._reflect_on_association(attribute) }
```

Comment:
I think rather than fallback on doing a query, we could get the column name from the association, no?

Reply:
This is probably unusual, but an association can also be a collection association like `has_many`. 
Should we handle `belongs_to` specially or better left as is now?

Reply:
> an association can also be a collection association like has_many.

Does that even work? I'd assume any association here would be a `belongs_to`?

Reply:
Yes, it works, just tested out in the console. 

https://github.com/rails/rails/blob/fb721b5d674ad5313ba8a835d4b87f1f204a2baf/activerecord/lib/active_record/validations/uniqueness.rb#L85-L96

Reply:
Oh wow...

Ok, let's handle `belongs_to` specifically, and fallback for the rest.

Reply:
So, `has_many` relations work only for `:scope`, but something like `validates_uniqueness_of :comments` does not work.
Handling `belongs_to` specifically leads to some hairy code. I'm thinking, do we even need to care about `has_many`s in the `:scope`? While it works accidentally, I can't imagine a use case for this.

Reply:
>  I can't imagine a use case for this.

Yeah, me neither.